pkg/webhook/certs.go

```diff
@@ -0,0 +1,486 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"math/big"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"time"
+)
+
+const (
+	caName     = "gatekeeper-ca"
+	namespace  = "gatekeeper-system"
+	service    = "gatekeeper-webhook-service"
+	certName   = "tls.crt"
+	keyName    = "tls.key"
+	caCertName = "ca.crt"
+	caKeyName  = "ca.key"
+	rotationCheckFrequency = 12 * time.Hour
+	certValidityDuration = 10 * 365 * 24 * time.Hour
+)
+
+var crLog = logf.Log.WithName("cert-rotation")
+
+var (
+	secretKey = types.NamespacedName{
+		Namespace: namespace,
+		Name:      "gatekeeper-webhook-server-cert",
+	}
+	// DNS name is <service name>.<namespace>.svc
+	DNSName = fmt.Sprintf("%s.%s.svc", service, namespace)
+	vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
+	vwhKey = types.NamespacedName{Name: "gatekeeper-validating-webhook-configuration"}
+)
+
+var _ manager.Runnable = &certRotator{}
+
+func AddRotator(mgr manager.Manager) (error) {
+	// Use a new client so we are unaffected by the cache sync kill signal
+	cli, err := client.New(mgr.GetConfig(), client.Options{Scheme: mgr.GetScheme(), Mapper: mgr.GetRESTMapper()})
+	if err != nil {
+		return err
+	}
+	rotator := &certRotator{client: cli}
+	mgr.Add(rotator)
+
+	reconciler := &ReconcileVWH{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+	if err := addController(mgr, reconciler); err != nil {
+		return err
+	}
+	return nil
+}
+
+type certRotator struct {
+	client client.Client
+}
+
+func (cr *certRotator) Start(stop <-chan (struct{})) error {
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if restart, err := cr.refreshCertIfNeeded(); err != nil {
+		crLog.Error(err, "could not refresh cert on startup")
+		return errors.Wrap(err, "could not refresh cert on startup")
+	} else if restart {
+		crLog.Info("certs refreshed, restarting server")
+		return nil
+	}
+	ticker := time.NewTicker(rotationCheckFrequency)
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if restart, err := cr.refreshCertIfNeeded(); err != nil {
+					crLog.Error(err, "error rotating certs")
+				} else if restart {
+					crLog.Info("certs refreshed, restarting server")
+					close(done)
+					return
+				}
+			case <-stop:
+				close(done)
+				return
+			}
+		}
+	}()
+
+	<-done
```

Comment:
Since you are blocking here anyways, the goroutine and done channel above are actually unnecessary. Just run the for loop and break on case <-stop .